crates/block-producer/src/mempool/dependency_graph.rs

```diff
@@ -0,0 +1,826 @@
+                .map(|children| children.difference(&processed))
+                .into_iter()
+                .flatten()
+                .cloned();
+
+            to_process.extend(unprocessed_children);
+
+            processed.insert(key);
+        }
+
+        // Only the original keys and the current roots need to be considered as roots.
+        //
+        // The children of the input keys are disqualified by definition (they're descendents),
+        // and current roots must be re-evaluated since their parents may have been requeued.
+        std::mem::take(&mut self.roots)
+            .into_iter()
+            .chain(keys)
+            .for_each(|key| self.try_make_root(key));
+
+        Ok(())
+    }
+
+    /// Removes a set of previously processed nodes from the graph.
+    ///
+    /// This is used to bound the size of the graph by removing nodes once they are no longer
+    /// required.
+    ///
+    /// # Errors
+    ///
+    /// Errors if
+    ///   - any node is unknown
+    ///   - any node is __not__ processed
+    ///   - any parent node would be left unpruned
+    ///
+    /// The last point implies that all parents of the given nodes must either be part of the set,
+    /// or already been pruned.
+    ///
+    /// This method is atomic.
+    pub fn prune_processed(&mut self, keys: BTreeSet<K>) -> Result<Vec<V>, GraphError<K>> {
+        let missing_nodes = keys
+            .iter()
+            .filter(|key| !self.vertices.contains_key(key))
+            .cloned()
+            .collect::<BTreeSet<_>>();
+        if !missing_nodes.is_empty() {
+            return Err(GraphError::UnknownNodes(missing_nodes));
+        }
+
+        let unprocessed = keys.difference(&self.processed).cloned().collect::<BTreeSet<_>>();
+        if !unprocessed.is_empty() {
+            return Err(GraphError::UnprocessedNodes(unprocessed));
+        }
+
+        // No parent may be left dangling i.e. all parents must be part of this prune set.
+        let dangling = keys
+            .iter()
+            .flat_map(|key| self.parents.get(key))
+            .flatten()
+            .filter(|parent| !keys.contains(parent))
+            .cloned()
+            .collect::<BTreeSet<_>>();
+        if !dangling.is_empty() {
+            return Err(GraphError::DanglingNodes(dangling));
+        }
+
+        let mut pruned = Vec::with_capacity(keys.len());
+
+        for key in keys {
+            let value = self.vertices.remove(&key).expect("Checked in precondition");
+            pruned.push(value);
+            self.processed.remove(&key);
+            self.parents.remove(&key);
+
+            let children = self.children.remove(&key).unwrap_or_default();
+
+            // Remove edges from children to this node.
+            for child in children {
+                if let Some(child) = self.parents.get_mut(&child) {
+                    child.remove(&key);
+                }
+            }
+        }
+
+        Ok(pruned)
```

Comment:
`pruned` and `keys` parameter should contain the same data, right? We should probably describe the returned value in the doc comments for this function.

But also, why do we need to return the same data as was passed in?

Reply:
> `pruned` and `keys` parameter should contain the same data, right? We should probably describe the returned value in the doc comments for this function.

I'll document this; though I'm inclined to make the change below instead which will nullify the need for this return value. Would do this in a separate PR ofc.

> But also, why do we need to return the same data as was passed in?

The method returns the value `V` for each key `K` passed in. The usage of this is in `Mempool::block_committed` and flows something like:

1. Commit batches -> returns `[V: TransactionId]`, which is used to
2. commit transactions -> returns `[V: AuthenticatedTransaction]`, which is used to
3. commit state in `InflightState` which uses the tx data to create the state deltas

It might make more sense to have `InflightState` keep a mapping of `TransactionId -> Delta` internally, then committing it only requires the IDs and not the full data. This makes sense to me from the perspective of data ownership, as it means `InflightState` doesn't depend on the caller to maintain the data.